language-extensions/dotnet-core-CSharp/src/managed/CSharpInputDataSet.cs

```diff
@@ -129,7 +129,25 @@ private unsafe void AddColumn(
                 case SqlDataType.DotNetChar:
                     int[] strLens = new int[rowsNumber];
                     Interop.Copy((int*)colMap, strLens, 0, (int)rowsNumber);
-                    CSharpDataFrame.Columns.Add(new StringDataFrameColumn(_columns[columnNumber].Name, DataSetUtils.StringSplitToArray(Interop.UTF8PtrToStr((char*)colData), strLens)));
+
+                    // Use byte-based splitting to properly handle multi-byte UTF-8 characters
+                    //
+                    CSharpDataFrame.Columns.Add(new StringDataFrameColumn(_columns[columnNumber].Name, DataSetUtils.UTF8ByteSplitToArray((byte*)colData, strLens)));
+                    break;
+                case SqlDataType.DotNetWChar:
+                    int[] wcharStrLens = new int[rowsNumber];
+                    Interop.Copy((int*)colMap, wcharStrLens, 0, (int)rowsNumber);
+    
+                    // For NCHAR/WCHAR, the strLenOrNullMap contains byte lengths, but we need character counts for UTF16PtrToStr
+                    // Each Unicode character is 2 bytes (sizeof(wchar_t)), so divide by 2 to get character count
+                    //
```

Comment:
on Linux, sizeof wchar is 4 bytes as opposed to 2bytes on windows. So, please be careful when calculating chars. I know the extension is not implemented for Linux, but we should not add to our debt to handle linux explicitly